packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js

```diff
@@ -0,0 +1,269 @@
+        case 'pending':
+          throw record.value;
+        case 'rejected':
+          throw record.value;
+        case 'resolved':
+          return record.value;
+      }
+    } else {
+      const thenable = {
+        pings: [],
+        then(resolve) {
+          if (newRecord.status === 'pending') {
+            thenable.pings.push(resolve);
+          } else {
+            Promise.resolve().then(() => resolve(newRecord.value));
+          }
+        },
+      };
+
+      const newRecord = {
+        status: 'pending',
+        value: thenable,
+      };
+      textCache.set(text, newRecord);
+
+      throw thenable;
+    }
+  }
+
+  function Text({text}) {
+    return text;
+  }
+
+  function AsyncText({text}) {
+    return readText(text);
+  }
+
+  // @gate experimental
+  it('should asynchronously load the suspense boundary', async () => {
+    await act(async () => {
+      ReactDOMFizzServer.pipeToNodeWritable(
+        <div>
+          <Suspense fallback={<Text text="Loading..." />}>
+            <AsyncText text="Hello World" />
+          </Suspense>
+        </div>,
+        writable,
+      );
+    });
+    expect(getVisibleChildren(document.body)).toEqual(<div>Loading...</div>);
+    await act(async () => {
+      resolveText('Hello World');
+    });
+    expect(getVisibleChildren(document.body)).toEqual(<div>Hello World</div>);
+  });
+
+  // @gate experimental
+  it('waits for pending content to come in from the server and then hydrates it', async () => {
+    const ref = React.createRef();
+
+    function App() {
+      return (
+        <div>
+          <Suspense fallback="Loading...">
+            <h1 ref={ref}>
+              <AsyncText text="Hello" />
+            </h1>
+          </Suspense>
+        </div>
+      );
+    }
+
+    await act(async () => {
+      ReactDOMFizzServer.pipeToNodeWritable(
+        // We currently have to wrap the server node in a container because
+        // otherwise the Fizz nodes get deleted during hydration.
+        <div id="container">
+          <App />
+        </div>,
+        writable,
+      );
+    });
+
+    // We're still showing a fallback.
+
+    // Attempt to hydrate the content.
+    const container = document.body.firstChild;
+    const root = ReactDOM.createRoot(container, {hydrate: true});
+    root.render(<App />);
+    Scheduler.unstable_flushAll();
+
+    // We're still loading because we're waiting for the server to stream more content.
+    expect(getVisibleChildren(container)).toEqual(<div>Loading...</div>);
+
+    // The server now updates the content in place in the fallback.
+    await act(async () => {
+      resolveText('Hello');
+    });
+
+    // The final HTML is now in place.
+    expect(getVisibleChildren(container)).toEqual(
+      <div>
+        <h1>Hello</h1>
+      </div>,
+    );
+    const h1 = container.getElementsByTagName('h1')[0];
+
+    // But it is not yet hydrated.
+    expect(ref.current).toBe(null);
+
+    Scheduler.unstable_flushAll();
```

Comment:
Maybe you could have two acts: one for Fizz, one for the Scheduler/reconciler?

Reply:
I was kind of expecting to have them all be like one act that does both the server and the client in that act. E.g. I didn't include anything like "scheduler.yield" here.

This was ported from another test though. Not sure if this will come up more.